File: synthetic/script/algorithm/cancer.py
from math import ceil
from algorithm.algorithm import Algorithm
from base.controller import Controller
from base.file_system import FileSystem
from base.configs import Configs
from utils.commands import Commands
import os
import re
import numpy as np
class Cancer(Algorithm):
    def __init__(self, controller:Controller) -> None:
        super().__init__()
        self.name = "cancer"
        self.__controller = controller
        self.__controller.addAlgorithm(self)

    def __calculateRank(self, u):
        return ceil(8/((1-u)/0.1))

    def __translateCancerPatterns(self, experiment_path): #pattern = [{}, {}, {}]
        dimensions = len(self.__controller.getParameter("dataset_size"))
        cancer_patterns = [] # [[{},{}], [{},{}]]
        lambda_0 = self.__controller.current_dataset.getDensity()

        experiment_folder = re.sub("/cancer.experiment", "", experiment_path)
        temp_folder = f"{experiment_folder}/temp"

        plain_patterns = os.listdir(temp_folder)
        for plain_pattern in plain_patterns:
            tuples = [set() for dimension in range(dimensions)]
            pattern_path = f"{temp_folder}/{plain_pattern}"
            
            data = ""
            with open(pattern_path) as pattern_file:
                for line in pattern_file:
                    data += line
            data = data.strip('[]')

            numpy_pattern = np.array(np.matrix(data))
            for index, value in np.ndenumerate(numpy_pattern):
                # if value < 0.5: # dont add dimensions to tuple
                if value <= lambda_0: # dont add dimensions to tuple
                    continue
                for n in range(len(index)): # iterates over all indices of the index
                    nth_tuple = tuples[n]
                    nth_tuple.add(index[n])

            cancer_patterns.append(tuples)

        return cancer_patterns

    def __createCancerFile(self):
        cancer_patterns = self.__translateCancerPatterns(self.experiment_path)

        with open(self.experiment_path, "a") as cancer_file:
            for pattern in cancer_patterns: #pattern = [{}, {}, {}]
                line = ""
                for d_tuple in pattern:
                    if len(d_tuple) == 0:
                        continue

                    line += str(d_tuple).replace("{","").replace("}","").replace(" ","") # d_tuple = {}
                    line += " "
                line = line.strip()
                if line == "":
                    print("CANCER found no patterns")
                    continue

                line += f" {self.__controller.current_dataset.calculateTuplesDensity(pattern):.6f}"
                line += "\n"
                cancer_file.write(line)

    def run(self, u, observations, timeout):
        if len(Configs.getParameter("dataset_size")) > 2:
            return True

        matlab_folder = ""
        # if self.__controller.ufmgMode():
            # matlab_folder = Configs.ufmgMatlabFolder()

        rank = Configs.getParameter("n_patterns")

        current_experiment = self.__controller.current_experiment
        current_iteration_folder = f"{self.__controller.current_iteration_folder}"

        translated_tensor_path = f"{current_iteration_folder}/tensors/mat/dataset-co{observations}.mat"
        self.experiment_path = f"{current_iteration_folder}/output/{current_experiment}/experiments/cancer.experiment"
        temp_folder = f"{current_iteration_folder}/output/{current_experiment}/experiments/temp"
        self.log_path = f"{current_iteration_folder}/output/{current_experiment}/logs/cancer.log"
        
        command = f"/usr/bin/time -o {self.log_path} -f 'Memory (kb): %M' "
        command += f"{matlab_folder}matlab -nodisplay -r 'cd(\"algorithm\"); "
        command += f"cancer({rank},\"../{translated_tensor_path}\","
        command += f"\"../{current_iteration_folder}\","
        command += f"\"{current_experiment}\");exit' | tail -n +11 "

        print(command)
        timedout = Commands.executeWithTimeout(command, timeout)     

        if timedout is False:
            self.__createCancerFile()
            
        FileSystem.delete(temp_folder)
        return timedout

        

File: synthetic/script/algorithm/cancer.m
% matlab -nodisplay -r 'cd("."); cancer(8,"../../iterations/1/tensors/plain/matrix9.txt","../../iterations/1","co1-u0.6");exit' | tail -n +11
function [] = cancer(rank, mat_path, current_iteration_folder, current_experiment)
addpath('../../algorithms/cancer/src/Product')
addpath('../../algorithms/cancer/src/Metrics')
addpath('../../algorithms/cancer/src/Cancer')

matrix = load(mat_path).matrix;
% matrix = str2num(matrix);

tic
[B, C] = Cancer(matrix, rank);
elapsed_time = toc;

temp_folder = current_iteration_folder + "/output/" + current_experiment + "/experiments/temp";
mkdir(temp_folder);

pattern_nb = 0;
% for i = 1 : size(matrix, 1)
for i = 1 : rank
    B_ = B(:,i);
    C_ = C(i,:);
    pattern = MaxProduct(B_, C_);
 
    if sum(abs(pattern(:))) > 0 
        pattern_nb = pattern_nb + 1;
        pattern_file_path = temp_folder + "/pattern_" + i + ".txt";
        file_id = fopen(pattern_file_path, "w");
        pattern = mat2str(pattern);
        fprintf(file_id, pattern);
    end
end

log_path = current_iteration_folder + "/output/" + current_experiment + "/logs/cancer.log";
file_id = fopen(log_path, "w");
fprintf(file_id, "filler:0\n");
fprintf(file_id, "filler:0\n");
fprintf(file_id, "filler:0\n");

pattern_nb = "Nb of patterns: " + pattern_nb + "\n";
fprintf(file_id, pattern_nb);

elapsed_time = "Run time: " + elapsed_time + '\n';
fprintf(file_id, elapsed_time);
end
